addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/thermostat/NATherm1Handler.java

```diff
@@ -89,23 +110,98 @@ protected State getNAThingProperty(String channelId) {
+    private State getSetpoint() {
+        return module != null
+                ? module.getSetpoint() != null ? toStringType(module.getSetpoint().getSetpointMode()) : UnDefType.NULL
+                : UnDefType.UNDEF;
+    }
+
+    private State getCurrentSetpoint() {
+        if (module != null && module.getSetpoint() != null) {
+            NASetpoint setPoint = module.getSetpoint();
+            String currentMode = setPoint.getSetpointMode();
+
+            NAThermProgram currentProgram = module.getThermProgramList().stream().filter(p -> p.getSelected() != null)
+                    .findFirst().get();
+            if (currentProgram != null) {
+                switch (currentMode) {
+                    case CHANNEL_SETPOINT_MODE_MANUAL:
+                        return toDecimalType(setPoint.getSetpointTemp());
+                    case CHANNEL_SETPOINT_MODE_AWAY:
+                        NAZone zone = getZone(currentProgram.getZones(), 2);
+                        return toDecimalType(zone.getTemp());
+                    case CHANNEL_SETPOINT_MODE_HG:
+                        NAZone zone1 = getZone(currentProgram.getZones(), 3);
+                        return toDecimalType(zone1.getTemp());
+                    case CHANNEL_SETPOINT_MODE_PROGRAM:
+                        NATimeTableItem currentProgramMode = getCurrentProgramMode(module.getThermProgramList());
+                        if (currentProgramMode != null) {
+                            NAZone zone2 = getZone(currentProgram.getZones(), currentProgramMode.getId());
+                            return toDecimalType(zone2.getTemp());
+                        }
+                    case CHANNEL_SETPOINT_MODE_OFF:
+                    case CHANNEL_SETPOINT_MODE_MAX:
+                        return UnDefType.NULL;
+                }
+            }
+        }
+        return UnDefType.UNDEF;
+    }
+
+    private NAZone getZone(List<NAZone> zones, int searchedId) {
+        NAZone result = zones.stream().filter(z -> z.getId() == searchedId).findFirst().get();
+
+        return result;
+    }
+
+    private long getNetatmoProgramBaseTime() {
+
+        // private int getNextSchedule(List<NAThermProgram> thermProgramList) {
         Calendar mondayZero = Calendar.getInstance();
         mondayZero.set(Calendar.DAY_OF_WEEK, Calendar.MONDAY);
         mondayZero.set(Calendar.HOUR_OF_DAY, 0);
         mondayZero.set(Calendar.MINUTE, 0);
         mondayZero.set(Calendar.SECOND, 0);
+        return mondayZero.getTimeInMillis();
+    }
+
+    private NATimeTableItem getCurrentProgramMode(List<NAThermProgram> thermProgramList) {
+        NATimeTableItem lastProgram = null;
+        Calendar now = Calendar.getInstance();
+        long diff = (now.getTimeInMillis() - getNetatmoProgramBaseTime()) / 1000 / 60;
+
+        NAThermProgram currentProgram = thermProgramList.stream()
+                .filter(p -> p.getSelected() != null && p.getSelected()).findFirst().get();
```

Comment:
FATAL: Note that `findFirst` will return an optional, if you call get on an empty optional it will throw an exception instead of returning `null`, so the check at line 192 does not make sense. Instead you should store the `Optional` and use `!isEmpty` on it.